File: common/algorithm/window.lua
-- KEYS[1]: 限流key
-- ARGV[1]: 当前时间戳(ms)
-- ARGV[2]: 窗口大小(ms)
-- ARGV[3]: 阈值
-- ARGV[4]: 本次消耗配额

local key = KEYS[1]
local now = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local threshold = tonumber(ARGV[3])
local cost = tonumber(ARGV[4])

-- 1. 移除窗口外的数据
local windowStart = now - window
redis.call('ZREMRANGEBYSCORE', key, '-inf', windowStart)

-- 2. 获取窗口内当前计数
local current = redis.call('ZCARD', key)

-- 3. 判断是否允许通过
local allowed = (current + cost) <= threshold
local remaining = threshold - current

if allowed then
    -- 4. 添加当前请求到集合
    local member = now .. ':' .. redis.call('INCR', key .. ':seq')
    redis.call('ZADD', key, now, member)
    remaining = remaining - cost
end

-- 5. 返回结构化结果
return {
    allowed and 1 or 0,  -- allowed
    remaining            -- remaining
}

File: common/algorithm/window.go
package algorithm

import (
	"context"
	_ "embed"
	"fmt"
	"github.com/zeromicro/go-zero/core/stores/redis"
)

//go:embed window.lua
var slideWindowScript string

// GetSlideWindowScript 获取滑动窗口Lua脚本
func GetSlideWindowScript() string {
	return slideWindowScript
}

// EvalSlideWindow 执行滑动窗口限流判定
// key: Redis限流计数器key
// nowMs: 当前时间戳(毫秒)
// windowMs: 窗口大小(毫秒)
// threshold: 阈值
// cost: 本次请求消耗的配额数
func EvalSlideWindow(ctx context.Context, redis *redis.Redis, key string, nowMs, windowMs, threshold, cost int64) (allowed bool, remaining int64, err error) {
	if redis == nil {
		return false, 0, fmt.Errorf("redis client is nil")
	}
	if key == "" {
		return false, 0, fmt.Errorf("key is empty")
	}
	if windowMs <= 0 {
		return false, 0, fmt.Errorf("invalid window size: %d", windowMs)
	}
	if threshold <= 0 {
		return false, 0, fmt.Errorf("invalid threshold: %d", threshold)
	}
	if cost <= 0 {
		return false, 0, fmt.Errorf("invalid cost: %d", cost)
	}

	// 执行Lua脚本
	resp, err := redis.EvalCtx(ctx, slideWindowScript, []string{key}, []interface{}{nowMs, windowMs, threshold, cost})
	if err != nil {
		return false, 0, fmt.Errorf("failed to eval script: %v", err)
	}

	// 解析返回值
	result, ok := resp.([]interface{})
	if !ok || len(result) != 2 {
		return false, 0, fmt.Errorf("invalid script response format")
	}

	// 转换返回值
	allowedInt, ok := result[0].(int64)
	if !ok {
		return false, 0, fmt.Errorf("invalid allowed value type")
	}
	remaining, ok = result[1].(int64)
	if !ok {
		return false, 0, fmt.Errorf("invalid remaining value type")
	}

	return allowedInt == 1, remaining, nil
}

// PreloadScript 预加载Lua脚本到Redis并返回SHA
func PreloadScript(redis *redis.Redis) (string, error) {
	if redis == nil {
		return "", fmt.Errorf("redis client is nil")
	}
	return redis.ScriptLoad(slideWindowScript)
}
